mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp

```diff
@@ -2848,31 +2848,75 @@ struct LLVMOpAsmDialectInterface : public OpAsmDialectInterface {
 // DialectInlinerInterface
 //===----------------------------------------------------------------------===//
 
+// Check whether the given alloca is an input to a lifetime intrinsic,
+// optionally passing through one or more casts on the way.
+static bool hasLifetimeMarkers(LLVM::AllocaOp allocaOp) {
+  SmallVector<Operation *, 2> stack(allocaOp->getUsers().begin(),
+                                    allocaOp->getUsers().end());
+  while (!stack.empty()) {
+    Operation *op = stack.pop_back_val();
+    if (isa<LLVM::LifetimeStartOp, LLVM::LifetimeEndOp>(op))
+      return true;
+    if (isa<LLVM::BitcastOp>(op))
+      stack.append(op->getUsers().begin(), op->getUsers().end());
+  }
```

Comment:
Do we need to handle the case where an allocated value is handed on per branch argument?

I assume it is not needed since we are only working on alloca operations placed in the entry block so they should be visible in all blocks of the function. Canonicalization should thus drop any alloca values passed via block arguments...

Reply:
You cannot assume that cannonicalization was executed before this pass, so I would assume that it's required.